From a17d7c1ff24f3b5d11430eb1f99d60722752e1f9 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Tue, 7 Nov 2023 12:10:36 +0100
Subject: [PATCH 0063/1695] plat-stm32mp1: pwr: use IO_READ32_POLL_TIMEOUT()

Update stm32mp1_pwr driver to use IO_READ32_POLL_TIMEOUT() macro.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 5ab96557525ee899f2c426d7e41489e6ce50713a plat-stm32mp1: pwr: use IO_READ32_POLL_TIMEOUT()]
Change-Id: I3756e116643378feeb654647aa7a710c48e79437
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
[etienne: cherry-picked from 83b3f58769c57f34e7967ac8062a9c521526a6ee]
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.c | 12 +++++-------
 1 file changed, 5 insertions(+), 7 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.c
index 547cccdcd..e06e2a2b1 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.c
@@ -105,19 +105,17 @@ static TEE_Result stm32mp1_pwr_regu_set_state(struct regulator *regu,
 {
 	const struct pwr_regu_desc *desc = regu->priv;
 	uintptr_t cr3 = stm32_pwr_base() + PWR_CR3_OFF;
-	uint64_t to = 0;
 
 	assert(desc);
 
 	if (enable) {
-		io_setbits32_stm32shregs(cr3, desc->cr3_enable_mask);
+		uint32_t value = 0;
 
-		to = timeout_init_us(TIMEOUT_US_10MS);
-		while (!timeout_elapsed(to))
-			if (io_read32(cr3) & desc->cr3_ready_mask)
-				break;
+		io_setbits32_stm32shregs(cr3, desc->cr3_enable_mask);
 
-		if (!(io_read32(cr3) & desc->cr3_ready_mask))
+		if (IO_READ32_POLL_TIMEOUT(cr3, value,
+					   value & desc->cr3_ready_mask,
+					   0, TIMEOUT_US_10MS))
 			return TEE_ERROR_GENERIC;
 	} else {
 		io_clrbits32_stm32shregs(cr3, desc->cr3_enable_mask);
-- 
2.47.2

